lib/grammers-client/src/client/auth.rs

```diff
@@ -251,20 +281,128 @@ impl Client {
         {
             Ok(tl::enums::auth::Authorization::Authorization(x)) => {
                 // Safe to unwrap, Telegram won't send `UserEmpty` here.
-                Ok(x.user.try_into().unwrap())
+                Ok(SignInResult::SignedIn(x.user.try_into().unwrap()))
             }
             Ok(tl::enums::auth::Authorization::SignUpRequired(x)) => {
-                Err(SignInError::SignUpRequired {
+                Ok(SignInResult::SignUpRequired {
                     terms_of_service: x.terms_of_service.map(|tos| tos.into()),
                 })
             }
+            Err(InvocationError::Rpc(RpcError { name, .. }))
+                if name == "SESSION_PASSWORD_NEEDED" =>
+            {
+                let password_information = self.get_password_information().await?;
+                Ok(SignInResult::PasswordRequired {
+                    password_information,
+                })
+            }
             Err(InvocationError::Rpc(RpcError { name, .. })) if name.starts_with("PHONE_CODE_") => {
                 Err(SignInError::InvalidCode)
             }
             Err(error) => Err(SignInError::Other(error)),
         }
     }
 
+    /// Extract information needed for the two-factor authentication
+    /// It's called automatically when we get SESSION_PASSWORD_NEEDED error during sign in.
+    pub async fn get_password_information(
```

Comment:
Do we want this method to be public? Why is the `Err` variant a `SignInError`, shouldn't it simply be `InvocationError`?

Reply:
Agree about `SignInError`.

This function is public because I wanted to provide a way for case telegram sends incorrect data with a `password` object. There is a verification for `p` and `g` parameters and docs say that we should perform it. I don't think that telegram can send incorrect values, but still...
Okay, another opinion. I make this function public and call it again if the parameters are incorrect to get the new params. We can call it for the new params two times more and panic with `telegram is broken, go use whatapp` (kidding) in case tg will continue to send incorrect params. What do you think?

Reply:
Ah yeah I forgot to mention this earlier. If the parameters are invalid, I think the library should retry in-place once, and if that fails again, then we return invocation failed. Users can't really do much about this, and fetching the required parameters should probably not be responsibility of the user (why should they care how 2FA is handled?). Then this method can become private.

Reply:
> then we return invocation failed.

The method returns `SignInError`. Invocation error can be wrapped by `SignInError::Other`, but it doesn't seem nice to return `SESSION_PASSWORD_NEEDED`. Looks like we need either 1) Create a new invocation error with `failed` or 2) Create a new `SignInError` or 3) Panic. TBH, none of this approaches looks good to me. What do you think about it?

Reply:
Let's `panic!()`, since this might indicate a bug somewhere.